src/main/java/org/dependencytrack/tasks/metrics/VulnerabilityMetricsUpdateTask.java

```diff
@@ -59,104 +62,75 @@ private void updateMetrics() throws Exception {
-     *
-     * @param pm     The {@link  PersistenceManager} to use
-     * @param lastId Highest ID of the previously fetched page
-     * @return Up to {@code 500} {@link Vulnerability} objects
-     * @throws Exception If the query could not be closed
-     */
-    private List<Vulnerability> fetchNextVulnerabilitiesPage(final PersistenceManager pm, final Long lastId) throws Exception {
-        try (final Query<Vulnerability> query = pm.newQuery(Vulnerability.class)) {
-            if (lastId != null) {
-                query.setFilter("id < :lastId");
-                query.setParameters(lastId);
-            }
-            query.setOrdering("id DESC");
-            query.range(0, 500);
-            query.getFetchPlan().setGroup(Vulnerability.FetchGroup.METRICS_UPDATE.name());
-            return List.copyOf(query.executeList());
+            /**
+             * 
+             * The created field has priotiy over the published field, which is used as a fallback
+             * However, the created field is always empty (in my instances)
+             * But we leave this mechanism and field juggling in place for backwards compatibility,
+             * and for (future) analyzers/sources that might provide this field.
+             * 
+             * BTW the queries to get these vulnerability counts are very fast,
+             * so strictly speaking there is no reason to create this extra table with metrics
+             *  
+             */
+
+            // Get metrics by published date but only if created field is null
+            Collection<YearMonthMetric> published = queryForMetrics(pm, "published", "created");
+
+            // Get metrics by created date regardless of value of published field
+            Collection<YearMonthMetric> created = queryForMetrics(pm, "created", null);
+
+            // Merge flat lists
+            published.addAll(created);
+
+            // Collect into nested map so we can sum the counts
+            Map<Object, Map<Object, Long>> metrics = published.stream()
+            .filter(ymm -> ymm.year != null)
+            .collect(Collectors.groupingBy(ymm -> ymm.year, 
+                            Collectors.groupingBy(ymm -> ymm.month, Collectors.summingLong(ymm -> ymm.count))));         
+
+            // Flatten again, but now into VulnerabilityMetrics that can be persisted
+            Stream<VulnerabilityMetrics> monthlyStream = metrics.entrySet().stream()
+            .flatMap(e -> e.getValue().entrySet().stream().flatMap(
+                v -> Stream.of(new VulnerabilityMetrics((Integer)e.getKey(), (Integer)v.getKey(), v.getValue().intValue(), measuredAt))));
+
+            // Flatten another time, for the yearly counts
+            Stream<VulnerabilityMetrics> yearlyStream = metrics.entrySet().stream()
+            .flatMap(e -> Stream.of(new VulnerabilityMetrics((Integer)e.getKey(), null, e.getValue().values().stream().mapToInt(d->d.intValue()).sum(), measuredAt)));
```

Comment:
You don't need `flatMap` (and thus also `Stream.of`) here, as you're only ever returning one element. So a `map` without the `Stream.of` part should suffice.

Reply:
ok, thanks